rfd/0048-desktop-access-session-recording.md

```diff
@@ -0,0 +1,268 @@
+
+    // Message is the encoded TDP message. It is not marshaled to JSON format
+    bytes Message = 2 [ (gogoproto.nullable) = true, (gogoproto.jsontag) = "-" ];
+
+    // DelayMilliseconds is the delay in milliseconds from the start of the session
+    int64 DelayMilliseconds = 3 [ (gogoproto.jsontag) = "ms" ];
+}
+```
+
+The `Message` field contains an artibrary TDP-encoded message, as defined in
+[RFD 0037](./0037-desktop-access-protocol.md). The vast majority of these events
+will contain TDP message 2 (PNG frame), but we will also capture:
+
+- message 1 - client screen spec: to know when the canvas needs to be resized
+  during playback
+- message 6 - clipboard data: to indicate when the clipboard was used in a
+  session
+- message 3 - mouse move
+- message 4 - mouse button
+
+Of note: we will not be capturing other user input events (keyboard or mouse
+wheel/scroll), as they are not necessary for playback. The result of keyboard
+input or scrolling captured in the session's PNG frames instead.
+
+We do, however, need to capture mouse movement and button clicks, as the RDP
+bitmaps sent from the Windows desktop do _not_ include the mouse cursor
+(RDP clients show the native OS cursor and send its position to the desktop).
+
+When this new event is introduced, several parts of the code that assume
+`SessionPrintEvent` is the only type of session recording will need to be
+udpated:
+
+- `AuditWriter` assumes every `[]byte` it receives needs to be packaged up into
+  a `SessionPrintEvent`. We will extend `AuditWriterConfig` with a new field:
+  `MakeEvent func(b []byte) AuditEvent`. This will enable SSH sessions to
+  configure `AuditWriter` to produce `SessionPrintEvent`s, while allowing
+  desktop sessions to emit `DesktopRecordingEvent`s.
+- `AuditWriter` does some tracking of `lastPrintEvent` in order to update
+  timestamps and chunk indicdes. This needs to be generalized to be able to
+  track the last event in desktop mode as well.
+- The `TeeStreamer` that filters out `SessionPrintEvent`s needs to be updated to
+  also filter out `DesktopRecordingEvent`s, as there will be a large number of
+  these events and we don't want to clutter the global audit log with them.
+
+SSH session recording can operate in synchronous or asynchronous mode, and can
+be captured at the node or at the proxy, resulting in a total of 4 possible
+configurations.
+
+Synchronous recording mode emits each audit event directly to the audit log, and
+fails if an event can not be written. Asynchronous recording mode writes events
+to a local filesystem log, and periodically uploads the audit events to external
+storage. Asynchronous mode is both more efficient (due to making less API calls)
+and more resilient (due to support for retries, backoff, stream recovery, etc).
+For desktop recordings, Teleport will support sync or async modes.
+
+Unlike SSH sessions, where the node itself is running Teleport and the user has
```

Comment:
I'm interested in feedback here. We do support sync/async modes, but we don't support record at node / record at proxy.

We could:

1. Leverage the existing `auth_service.session_recording` property, and:
  - both `node-sync` and `proxy-sync` mean: record in synchronous mode
  - both `node` and `proxy` mean: record in async mode
2. Create a new `desktop_session_recording` property that supports only these modes:
  - `sync`
  - `async`
  - `off`